### PR TITLE
new: upgrade nginx core to 1.25.1

### PR DESCRIPTION
**Pull request by swananan**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

https://github.com/openresty/openresty/pull/920

### CODE REVIEW

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

File: .travis.yml [hunk 1]
```diff
@@ -52,8 +52,8 @@ env:
     - TEST_NGINX_SLEEP=0.006
     - MALLOC_PERTURB_=9
   jobs:
-    - NGINX_VERSION=1.21.4 OPENSSL_VER=1.1.0l OPENSSL_PATCH_VER=1.1.0d
-    - NGINX_VERSION=1.21.4 OPENSSL_VER=1.1.1s OPENSSL_PATCH_VER=1.1.1f
+    - NGINX_VERSION=1.25.1 OPENSSL_VER=1.1.0l OPENSSL_PATCH_VER=1.1.0d
```
[the remark this answers is not in the document]
Comment (swananan) [reply .travis.yml:55]: sure

File: t/028-req-header.t [hunk 2]
```diff
@@ -274,7 +274,7 @@ Content-Type:
 --- request
 GET /bar
 --- response_body
-Foo: a
+Foo: a, b
```
[the remark this answers is not in the document]
Comment (swananan) [reply t/028-req-header.t:277]: based on this [https://github.com/nginx/nginx/commit/7dc6f4e25d21588249691aab8c6013c126eae258](nginx commit), the newer version of nginx combines the same $http_* variables together.

**Review by zhuizhuhaomeng — APPROVED** [review]

**Comment by splitice** [comment]
@swananan should http/3 be enabled without any tests?

Unless there's a pr planned shouldn't it be disabled and subsequently enabled when tests are available?

**Comment by swananan** [comment]
> 
Thanks for your reminder. Enabling the HTTP/3 by default is intended to test nginx core 1.25.1 with HTTP/3 feature using the existing lua-nginx-module test cases. Additionally, enabling HTTP/3 by default is achieved by adding a new CI build job to lua-nginx-module, which helps minimize the impact of the change. I also plan to undertake the following tasks: adding HTTP/3 support to test-nginx, add HTTP/3 test cases into lua-nginx-module, and find out or fix the features in lua-nginx-module which are not compatible with HTTP/3.

**Comment by kingluo** [comment]
> > 
> 
> I also plan to undertake the following tasks: adding HTTP/3 support to test-nginx, add HTTP/3 test cases into lua-nginx-module, and find out or fix the features in lua-nginx-module which are not compatible with HTTP/3.

@swananan 
FYI, how about taking a look at a simple test framework I wrote in bash? Use the curl quic version to send the request, and use grep and jq to parse the headers and responses.

https://github.com/kingluo/apisix/pull/5/files#diff-c68f158a2b36cdac767021d401314a595c446974727c24edc775469ea7646d5d

**Comment by swananan** [comment]
@kingluo Thanks for your information, which is very helpful. Since openresty use test-nginx as their test framework, and @zhuizhuhaomeng already have supported http3 in the test-nginx (also use the curl quic version), I guess I may not use this bash test framework directly, but I still benefit a lot from this.

**Comment by kingluo** [comment]
> @kingluo Thanks for your information, which is very helpful. Since openresty use test-nginx as their test framework, and @zhuizhuhaomeng already have supported http3 in the test-nginx (also use the curl quic version), I guess I may not use this bash test framework directly, but I still benefit a lot from this.

Yes, of course, test::nginx is the official mandatory testing framework.
The biggest advantage of the bash framework is that you can make full use of the options of each command, for example, you can use all the options of curl, and you can also use grep and jq filter expressions to inspect the response.
Hopefully, future versions of test::nginx will provide more flexibility.
BTW, it seems taht the current version of test::nginx still uses the `http3` directive keyword? While the Nginx official already changed it to `quic`.

**Comment by splitice** [comment]
I look forward to seeing these efforts move forward and are monitoring for areas to contribute.

Probably in the bug hunting area once an initial test framework and results are in.

**Comment by swananan** [comment]
> > @kingluo Thanks for your information, which is very helpful. Since openresty use test-nginx as their test framework, and @zhuizhuhaomeng already have supported http3 in the test-nginx (also use the curl quic version), I guess I may not use this bash test framework directly, but I still benefit a lot from this.
> 
> Yes, of course, test::nginx is the official mandatory testing framework. The biggest advantage of the bash framework is that you can make full use of the options of each command, for example, you can use all the options of curl, and you can also use grep and jq filter expressions to inspect the response. Hopefully, future versions of test::nginx will provide more flexibility. BTW, it seems that the current version of test::nginx still uses the `http3` keyword in listen directive? While the Nginx official already changed it to `quic`.

@kingluo Yes, test-nginx supported http3 in 2021, so it used the old http3 config, I guess it have been fixed and @zhuizhuhaomeng haven't updated his new changes in the test-nginx.

**Pull request merged**